tracer/src/Datadog.Trace/Logging/Internal/DatadogDeferredSink.cs

```diff
@@ -0,0 +1,95 @@
+﻿// <copyright file="DatadogDeferredSink.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using Datadog.Trace.Vendors.Serilog.Core;
+using Datadog.Trace.Vendors.Serilog.Events;
+using Datadog.Trace.Vendors.Serilog.Sinks.File;
+
+namespace Datadog.Trace.Logging;
+
+#if !NETCOREAPP
+internal sealed class DatadogDeferredSink : ILogEventSink, IFlushableFileSink, IDisposable
+#else
+internal sealed class DatadogDeferredSink : ILogEventSink, IFlushableFileSink, IDisposable, IThreadPoolWorkItem
+#endif
+{
+    private readonly ConcurrentQueue<LogEvent> _queue;
+    private readonly ILogEventSink _sink;
+#if !NETCOREAPP
+    private readonly WaitCallback _waitCallback;
+#endif
+    private long _active;
+
+    public DatadogDeferredSink(ILogEventSink sink)
+    {
+        _sink = sink;
+        _queue = new ConcurrentQueue<LogEvent>();
+#if !NETCOREAPP
+        _waitCallback = InternalWaitCallBackDelegate;
+#endif
+        _active = 0;
+    }
+
+    public void Emit(LogEvent logEvent)
+    {
+        _queue.Enqueue(logEvent);
+        if (Interlocked.CompareExchange(ref _active, 1, 0) == 0)
+        {
+#if !NETCOREAPP
+            ThreadPool.UnsafeQueueUserWorkItem(_waitCallback, null);
+#else
+            ThreadPool.UnsafeQueueUserWorkItem(this, true);
+#endif
```

Comment:
This stuff all makes me nervous, as I've basically not used it for 10 years, so I don't know the edge cases and issues, and it feels pretty high risk to be replacing a "core" component like this. 

It's also not entirely clear to me that it's safe to move the `LogEvent` to a background thread like this, rather than rendering it immediately. Is this something Serilog does internally? If not, why not? Have you checked if any of that has changed in Serilog 3.0?


So all I can really say is "I hope you've done this right" 😬

Reply:
>It's also not entirely clear to me that it's safe to move the LogEvent to a background thread like this, rather than rendering it immediately. Is this something Serilog does internally? If not, why not? Have you checked if any of that has changed in Serilog 3.0?

All this is done not because the rendering but for the [OpenFile handles of the sink](https://github.com/DataDog/dd-trace-dotnet/blob/master/tracer/src/Datadog.Trace/Vendors/Serilog.Sinks.File/Sinks/File/RollingFileSink.cs#L92-L112).

I wanted to defer all log handling when doing a cold start because basically we write a bunch of logs at startup and then we don't write any after the app is running normally.

In fact the whole PR is just an attempt  to reduce startup time , configuration load and log writes. I guess only CI Visibility and Serverless will benefits of this.

This is way this shouldn't affect Throughput tests, but Execution time benchmarks. Which in fact we do: https://github.com/DataDog/dd-trace-dotnet/pull/4360#issuecomment-1632212856

I'm open for ideas for the R&D week 😉